### PR TITLE
Add `toUnorderedList` to `PriorityQueue`.

### LINKED ISSUE #45: PriorityQueue and comparison function inconsistent
The documentat of `PriorityQueue` states:
```
  /// The [comparison] is a [Comparator] used to compare the priority of
  /// elements. An element that compares as less than another element has
  /// a higher priority.
```

(btw: the documentation doesn't state what the function should return).

However, the comparison function is also used as equality function.
For example, `PriorityQueue.contains`, or `PriorityQueue.remove` both use this function to determine equality. In some cases that works, but a common use case is to extract a priority field from one of many fields.

Example:
``` dart
class PriorityPoint {
  int x, y, priority;
  PriorityPoint(this.x, this.y, this.priority);
}

var queue = new PriorityQueue(comparison: (p1, p2) => p2.priority - p1.priority);
queue.add(new PriorityPoint(0, 1, 5));
queue.contains(new PriorityPoint(-1, -3, 5));  // Should probably return false.
```

I'm not sure what the best solution is, but the current behavior and documentation are not consistent.
Furthermore, the current behavior makes it hard to implement the provided use case.

### PR DESCRIPTION
**Pull request by lrhn**
Also change `HeapPriorityQueue` to use `==` for `contains` and `remove` checks.

Fix a bad null-safety migration for HeapPriorityQueue, and add tests to ensure it works with `null` values.

Fixes #45, #140

### CODE REVIEW

**Review by jakemac53 — COMMENTED** [review]
Is it actually useful to store nulls in a queue like this? That seems really dubious to me. We could instead make it `<E extends Object>`?

Maybe not worth restricting though...

File: lib/src/priority_queue.dart [hunk 1]
```diff
@@ -9,6 +9,18 @@ import 'utils.dart';
 /// A priority queue is a priority based work-list of elements.
 ///
 /// The queue allows adding elements, and removing them again in priority order.
+/// The same object can be added to the queue more than once.
+/// There is no specified oredering for objects with the same priority
```
Comment (natebosch) [inline lib/src/priority_queue.dart:13]:
```suggestion
/// There is no specified ordering for objects with the same priority
```

File: lib/src/priority_queue.dart [hunk 2]
```diff
@@ -223,18 +264,23 @@ class HeapPriorityQueue<E> implements PriorityQueue<E> {
...
     var set = SplayTreeSet<E>(comparison);
     for (var i = 0; i < _length; i++) {
-      set.add(_queue[i]!);
+      set.add(_elementAt(i));
     }
     return set;
   }
 
+  @override
+  List<E> toUnorderedList() => _toUnorderedList();
```
Comment (natebosch) [inline lib/src/priority_queue.dart:279]: Why do we need the private one? Can we put the implementation here and skip the function?

**Review by natebosch — APPROVED** [review]

File: lib/src/priority_queue.dart (hunk 2 above)
Comment (lrhn) [reply lib/src/priority_queue.dart:279]: I call this method from `toList` too. It's an implementation detail that the two methods share code, so I don't want to risk someone overriding `toUnorderedList` and also affecting `toList`.
Generally, I avoid calling an overridable method from another method unless the latter method is explicitly documented as depending on the former. Code sharing is an implementation detail, so it shouldn't affect the visible API.

**Comment by lrhn** [comment]
All our other collections accept `null`, so it would stand out if this one doesn't, and it would be a breaking change.
Using `as E` also works better than `!` with legacy code which might store `null`  values in a `HeapPriorityQueue<int>`, because `null as int*` doesn't throw, but `null!` does.

There is no real reason to not to accept `null` as a value. The `null` value is an object too.

The `_elementAt` code will only read a `null` if the `E` type actually accepts `null` - it's only used for indices in the range where we know we have stored elements. All the `as E` operations will only see `null` values actually stored into the heap, so it's safely compilable to JavaScript where we omit the `as` casts.

**Review by natebosch — APPROVED** [review]

**Review by jakemac53 — APPROVED** [review]

**Comment by mit-mit** [comment]
@lrhn ready to merge?

File: lib/src/priority_queue.dart [hunk 3]
```diff
@@ -267,10 +316,10 @@ class HeapPriorityQueue<E> implements PriorityQueue<E> {
     // in the heap will also have lower priority.
     do {
       var index = position - 1;
-      var element = _queue[index]!;
+      var element = _elementAt(index);
       var comp = comparison(element, object);
-      if (comp == 0) return index;
-      if (comp < 0) {
+      if (comp <= 0) {
+        if (comp == 0 && element == object) return index;
```
Comment (natebosch) [inline lib/src/priority_queue.dart:322]: This turned out to break at least one project's tests.

I think it's the right thing to do, and they were able to work around it.

Comment (lrhn) [reply lib/src/priority_queue.dart:322]: Super. If you rely on `compareTo` for equality, and don't override `==` to match, then I guess this would affect you.
